File: backend/routes/locations.js
const express = require('express');
const router = express.Router();

// In-memory store — replace with a DB in production
let locations = [
  { id: 'GH-NRB-001', name: 'Nairobi Central Hub',   status: 'online',  region: 'Nairobi' },
  { id: 'GH-MBA-042', name: 'Mombasa Coastal Link',   status: 'online',  region: 'Mombasa' },
  { id: 'GH-KIS-109', name: 'Kisumu West Station',    status: 'offline', region: 'Kisumu'  },
  { id: 'GH-ELD-215', name: 'Eldoret Tech Park',      status: 'online',  region: 'Eldoret' },
];

// GET /api/locations  — list all, enriched with live session data
router.get('/', (req, res) => {
  const routerSvc = require('../services/routerService');
  const connections = Array.from(routerSvc.activeConnections.values());
  const now = Date.now();

  const enriched = locations.map(loc => {
    // Count active sessions whose phone area code loosely maps to region
    // In production you'd store locationId on each session
    const activeUsers = connections.filter(c => new Date(c.expiryTime).getTime() > now).length;
    const revenue = connections
      .filter(c => new Date(c.expiryTime).getTime() > now)
      .reduce((s, c) => s + (parseFloat(c.packagePrice) || 0), 0);

    return {
      ...loc,
      activeUsers: loc.status === 'online' ? activeUsers : 0,
      dailyRevenue: loc.status === 'online' ? parseFloat((revenue / Math.max(locations.filter(l=>l.status==='online').length,1)).toFixed(2)) : 0,
    };
  });

  const online = enriched.filter(l => l.status === 'online');
  const totalActiveUsers = connections.filter(c => new Date(c.expiryTime).getTime() > now).length;
  const totalRevenue = connections.reduce((s, c) => s + (parseFloat(c.packagePrice) || 0), 0);

  res.json({
    locations: enriched,
    summary: {
      total: locations.length,
      online: online.length,
      offline: locations.length - online.length,
      activeUsers: totalActiveUsers,
      totalRevenue: totalRevenue.toFixed(2),
    }
  });
});

// POST /api/locations — add a new hotspot
router.post('/', (req, res) => {
  const { name, region } = req.body;
  if (!name || !region) {
    return res.status(400).json({ success: false, message: 'name and region are required' });
  }
  const id = 'GH-' + region.slice(0,3).toUpperCase() + '-' + String(Date.now()).slice(-3);
  const loc = { id, name, region, status: 'online' };
  locations.push(loc);
  console.log(`📍 New hotspot added: ${name} (${id})`);
  res.json({ success: true, location: loc });
});

// PATCH /api/locations/:id — update status
router.patch('/:id', (req, res) => {
  const loc = locations.find(l => l.id === req.params.id);
  if (!loc) return res.status(404).json({ success: false, message: 'Location not found' });
  if (req.body.status) loc.status = req.body.status;
  if (req.body.name) loc.name = req.body.name;
  res.json({ success: true, location: loc });
});

// DELETE /api/locations/:id
router.delete('/:id', (req, res) => {
  const idx = locations.findIndex(l => l.id === req.params.id);
  if (idx === -1) return res.status(404).json({ success: false, message: 'Location not found' });
  locations.splice(idx, 1);
  res.json({ success: true });
});

module.exports = router;

File: backend/handlers/locations.go
package handlers

import (
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/gornhom/backend/db"
	"github.com/gornhom/backend/models"
	"github.com/gornhom/backend/services"
)

type LocationHandler struct {
	rs *services.RouterService
}

var defaultLocations = []models.Location{
	{ID: "GH-NRB-001", Name: "Nairobi Central Hub", Status: "online", Region: "Nairobi"},
	{ID: "GH-MBA-042", Name: "Mombasa Coastal Link", Status: "online", Region: "Mombasa"},
	{ID: "GH-KIS-109", Name: "Kisumu West Station", Status: "offline", Region: "Kisumu"},
	{ID: "GH-ELD-215", Name: "Eldoret Tech Park", Status: "online", Region: "Eldoret"},
}

func NewLocationHandler(rs *services.RouterService) *LocationHandler {
	if exists, _ := db.LocationExists(); !exists {
		for _, l := range defaultLocations {
			db.InsertLocation(l)
		}
	}
	return &LocationHandler{rs: rs}
}

func (h *LocationHandler) List(c *gin.Context) {
	locs, err := db.GetAllLocations()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "message": err.Error()})
		return
	}

	connections := h.rs.GetActiveConnections()
	now := time.Now()

	onlineCount := 0
	for _, l := range locs {
		if l.Status == "online" {
			onlineCount++
		}
	}

	totalActive := 0
	totalRevenue := 0.0
	for _, conn := range connections {
		expiry, _ := time.Parse(time.RFC3339, conn.ExpiryTime)
		if expiry.After(now) {
			totalActive++
			totalRevenue += conn.PackagePrice
		}
	}

	enriched := make([]map[string]interface{}, 0, len(locs))
	for _, loc := range locs {
		activeUsers := 0
		revenue := 0.0
		if loc.Status == "online" && onlineCount > 0 {
			activeUsers = totalActive / onlineCount
			revenue = totalRevenue / float64(onlineCount)
		}
		enriched = append(enriched, map[string]interface{}{
			"id":           loc.ID,
			"name":         loc.Name,
			"status":       loc.Status,
			"region":       loc.Region,
			"activeUsers":  activeUsers,
			"dailyRevenue": revenue,
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"locations": enriched,
		"summary": gin.H{
			"total":        len(locs),
			"online":       onlineCount,
			"offline":      len(locs) - onlineCount,
			"activeUsers":  totalActive,
			"totalRevenue": totalRevenue,
		},
	})
}

func (h *LocationHandler) Add(c *gin.Context) {
	var body struct {
		Name   string `json:"name"`
		Region string `json:"region"`
	}
	if err := c.ShouldBindJSON(&body); err != nil || body.Name == "" || body.Region == "" {
		c.JSON(http.StatusBadRequest, gin.H{"success": false, "message": "name and region are required"})
		return
	}
	region := body.Region
	if len(region) < 3 {
		region = region + strings.Repeat("X", 3-len(region))
	}
	loc := models.Location{
		ID:     fmt.Sprintf("GH-%s-%d", strings.ToUpper(region[:3]), time.Now().UnixMilli()%1000),
		Name:   body.Name,
		Region: body.Region,
		Status: "online",
	}
	if err := db.InsertLocation(loc); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "message": err.Error()})
		return
	}
	c.JSON(http.StatusCreated, gin.H{"success": true, "location": loc})
}

func (h *LocationHandler) Update(c *gin.Context) {
	id := c.Param("id")
	var body struct {
		Status string `json:"status"`
		Name   string `json:"name"`
		Region string `json:"region"`
	}
	c.ShouldBindJSON(&body)
	if err := db.UpdateLocation(id, body.Status, body.Name, body.Region); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "message": err.Error()})
		return
	}
	locs, _ := db.GetAllLocations()
	for _, l := range locs {
		if l.ID == id {
			c.JSON(http.StatusOK, gin.H{"success": true, "location": l})
			return
		}
	}
	c.JSON(http.StatusNotFound, gin.H{"success": false, "message": "Location not found"})
}

func (h *LocationHandler) Delete(c *gin.Context) {
	id := c.Param("id")
	if err := db.DeleteLocation(id); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"success": false, "message": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"success": true})
}
